### DISCUSSION TITLE
Upgrading from Stork 1.1 to 1.2

### DISCUSSION THREAD

**Discussion by justinmayer** [jameslittle230/stork#166] [General]
Congratulations on the Stork 1.2 release! 🎉  After experimenting with it briefly, I have a few comments/suggestions:

* Maybe consider putting upgrade instructions at the bottom of the [release notes](https://github.com/jameslittle230/stork/releases/tag/v1.2.0) for each release? (Mastodon does this, and it's extremely helpful.)
* Perhaps update the documentation at the same time as project releases? For example, users might be confused when [the docs refer to a `filename` configuration directive](https://stork-search.net/docs/build), only to later see a deprecation notice when invoking `stork`.
* Running v1.2 of `stork` via [Invoke][] results in an error that never occurred in v1.1. Invoke debug output ends with:
```
[…]
invoke.executor.execute: Finished loading collection & shell env configs
invoke.program.run: Received a possibly-skippable exception: <UnexpectedExit: cmd='stork build --input ~/projects/test/search.toml --output ~/projects/test/search-index.st' exited=-2>
```

Can you think of what change(s) in the Stork 1.2 release might have caused that? I rely on [Invoke][] for project automation, so unfortunately I had to revert back to Stork 1.1.

Thanks for all the work you put into Stork. It’s a fantastic project! 😁

[Invoke]: https://github.com/pyinvoke/invoke

**Comment by jameslittle230** [comment]
Hi @justinmayer - Thanks for using Stork!

Those are both good points about the release process! Thanks for sending them along. I totally agree that upgrade instructions would be super helpful - I’ll add them (as well as publish a docs update) as soon as I can.

I’ve never heard of Invoke before, but I’m sorry to hear Stork 1.2.0 doesn’t play nicely with it. I can’t think of any incompatibilities off the top of my head, but I’ll play around with it and see if I can reproduce the errors and figure out what’s going on. In the meantime, you might try two things to narrow down the issue:
* If Invoke has a verbose mode, you could use it to hopefully get more output from Stork and from Invoke’s internals to see why the process is exiting
* I wonder if you’d see a similar failure if you had a very minimal configuration file. I don’t know what you’re indexing but if it’s huge, this could be an out-of-memory error (which would also be bad, but would let me know where to look to fix it).

Again, thanks for the kind words and for writing in! I really appreciate the comments and hope I can get Stork 1.2 working for you soon.

James

**Reply by justinmayer** [reply]
Many thanks for quick response, James. The Invoke output I included above was taken when running it with the `--debug` flag, which unfortunately I believe is as verbose as it gets.

The Stork configuration file in question contains entries to 14 rather pedestrian HTML files, which I imagine isn't a huge amount to index. Running `stork` v1.2 directly, outside of an Invoke task, works just fine.

**Reply by jameslittle230** [reply]
Interesting! I'll poke around with Invoke and see what I can find.

**Comment by jameslittle230** [comment]
I've figured it out!

I was able to create a minimally reproducible example, which you can download [here (invoke-test.zip)](https://github.com/jameslittle230/stork/files/6362924/invoke-test.zip). When I run 

```sh
$ invoke --echo --debug build
```

the process hangs forever. However, when I run

```sh
$ invoke --echo --debug --pty build
```

it works successfully. Let me know if you're seeing the same behavior.

I think a lot of the new command-line behavior I added in Stork 1.2.0 relies on accurate detection of stdout and stderr being TTYs. It seems like the `--pty` option tells the Stork process that it's being run in a pseudo-tty, which will allow it to complete successfully. You can read more on [this Invoke Q&A page](https://github.com/pyinvoke/invoke/blob/ed94c59f1eacc700dac3815530493e0809fe001d/sites/www/faq.rst#why-is-my-command-behaving-differently-under-invoke-versus-being-run-by-hand).

**Comment by jameslittle230** [comment]
I've also updated the documentation throughout the site to reflect the configuration options and command line API in 1.2.0 -- thanks for pushing me to do that :)

**Comment by justinmayer** [comment]
Bravo, James! Thank you *so* much for really diving into this and divining the source of the problem. I just tested this by appending `pty=True` to the task command definitions, and that did indeed fully resolve the problem. And in the process, I learned something about Invoke’s TTY handling. (I feel a bit silly for not thinking to check the Invoke FAQ, though! 😅)

Once again, thank you for going above and beyond. Truly much appreciated! 👏

**Reply by jameslittle230** [reply]
I'm really glad that worked! Please let me know if you have any other questions or feedback.

**Comment by justinmayer** [comment]
Hey James! I've done some more testing with Stork 1.2.1, and I've noticed a problem that I feel I can reliably reproduce.

If you download & expand the attached [`testsite.zip`](https://github.com/jameslittle230/stork/files/7294646/testsite.zip) file and follow the steps below, I'm hoping you will be able to reproduce it as well:

```shell
cd testsite
cp theme/js/stork-1.2.1.js theme/js/stork.js
cp theme/js/stork-1.2.1.wasm theme/js/stork.wasm
python3 -m http.server
# Navigate to http://localhost:8000, open a JS console, search for "markdown", and observe that it fails.
^C
cp theme/js/stork-1.1.wasm theme/js/stork.wasm
python3 -m http.server
# Navigate to http://localhost:8000, open a JS console, search for "markdown", and observe that it succeeds.
```

The following is representative of the errors I've seen when experimenting with v1.2.x of `stork.wasm`:

```
StorkError: Could not perform search: the WASM binary failed to return search results.
    You might not be serving your search index properly.
    If you think this is an error, please file a bug: https://jil.im/storkbug
    
    The WASM binary came back with:
    Index not found. You must parse an index before performing searches with it.
```
```js
t — storkError.ts:3
(anonymous function) — searchData.ts:59
(anonymous function) — entity.ts:167
(anonymous function) — entityDom.ts:362
inputInputEvent — entityDom.ts:110
(anonymous function) — entity.ts:176
(anonymous function) — entityDom.ts:362
inputInputEvent — entityDom.ts:110
entity.ts:176
```

Do you have any idea what might be done to resolve this problem?

**Reply by jameslittle230** [reply]
I'm going to track this in #200, because this might be a support issue that turns into a bug report.

I'm more than happy to help, but just for my own sanity going forward, please create new discussion threads or issues for distinct bugs so I don't lose things :)
